lib/srv/alpnproxy/kube.go

```diff
@@ -238,8 +241,20 @@ func (m *KubeMiddleware) reissueCertIfExpired(ctx context.Context, cert tls.Cert
 	}
 }
 
-// NewKubeListener creates a listener for kube local proxy.
-func NewKubeListener(casByTeleportCluster map[string]tls.Certificate) (net.Listener, error) {
+// NewKubeListenerWithRandomPort creates a listener for kube local proxy using a
+// random localhost port.
+func NewKubeListenerWithRandomPort(casByTeleportCluster map[string]tls.Certificate) (net.Listener, error) {
```

Comment:
I didn't have time to look into this closely, but why do kube proxies need a separate `*WithRandomPort` function whereas db proxies do not?

Reply:
db has a single ALPN local proxy, which should serve the port user specified. 

kube (and other HTTP proxies like aws/azure/gcp) has two local proxies, one HTTPS_PROXY (in code it's called forward proxy) and the ALPN local proxy. Client connects through HTTPS_PROXY so it should use the port user specified. The HTTPS_PROXY forwards to the ALPN local proxy where it can use a random port.

The preivous `NewKubeListener` creates the tcp listener within itself. But gateway code needs to pass in a tcp listener.